stan/math/gpu/matrix_gpu.hpp

```diff
@@ -158,6 +162,102 @@ class matrix_gpu {
     oclBuffer_ = a.buffer();
     return *this;
   }
+
+  /**
+   * Stores zeros in the matrix on the GPU.
+   * Supports writing zeroes to the lower and upper triangular or
+   * the whole matrix.
+   *
+   * @tparam triangular_view Specifies if zeros are assigned to
+   * the entire matrix, lower triangular or upper triangular with
+   * possible values stan::math:Entire, stan::math:Lower or
+   * stan::math::Upper
+   *
+   */
+  template <int triangular_view>
+  void zeros() {
+    if (size() == 0)
+      return;
+    cl::Kernel kernel = opencl_context.get_kernel("zeros");
+    cl::CommandQueue cmdQueue = opencl_context.queue();
+    try {
+      set_kernel_args(kernel, this->buffer(), this->rows(), this->cols(),
+                      triangular_view);
+      cmdQueue.enqueueNDRangeKernel(kernel, cl::NullRange,
+                                    cl::NDRange(this->rows(), this->cols()),
+                                    cl::NullRange, NULL, NULL);
+    } catch (const cl::Error& e) {
+      check_opencl_error("zeros", e);
+    }
+  }
+
+  /**
+   * Copies a lower/upper triangular of a matrix to it's upper/lower.
+   *
+   * @tparam triangular_map Specifies if the copy is
+   * lower-to-upper or upper-to-lower triangular with possible values
+   * stan:math::LowerToUpper and stan::math::UpperToLower
+   *
+   * @throw <code>std::invalid_argument</code> if the matrix is not square.
+   *
+   */
+  template <int triangular_map>
+  void triangular_transpose() {
+    if (size() == 0) {
+      return;
+    }
+    if (size() == 1) {
+      return;
+    }
+    check_size_match("triangular_transpose (GPU)",
```

Comment:
We have a `check_square()` function for just this reason. Can we change this to use `check_square`?

Reply:
The check_square expects `const matrix_gpu&` and `triangular_transpose` is inside this class. The check_square.hpp expects this class to be defined.

Reply:
I am not sure how to handle this.

Reply:
One thing we can do here is add a template to GPU `check_square()` so we can remove the dependency on `matrix_gpu`

```cpp
template <typename T>
inline void check_square(const char* function, const char* name,
                         const T& y)
```

This shouldn't clash with the `check_square()` in `prim/err` since the compiler should know we've explicitly defined a `check_square()` with an eigen matrix in `prim/err`.

@syclik does that make sense?